### PR TITLE
FC2

### PR DESCRIPTION
**Pull request by misgeatgit**
-  Added R2L specific forward chaining.
-  Changed the rule choosing algorithm.
-  Used nested atomspace for separating partially grounded rules form main atomspace inorder to prevent undesirable inferences.

### CODE REVIEW

File: opencog/rule-engine/forwardchainer/ForwardChainer.cc [hunk 1]
```diff
@@ -22,20 +22,26 @@
...
+#include "VarGroundingPMCB.h"
 
 using namespace opencog;
 
 ForwardChainer::ForwardChainer(AtomSpace& as, Handle rbs) :
-	_as(as), _rec(_as), _rbs(rbs), _configReader(as, rbs), _fcmem(&as)
+	_as(as), _rec(as), _rbs(rbs), _configReader(as, rbs), _fcmem(&as)
 {
-    init();
+   init();
```
Comment (williampma) [inline opencog/rule-engine/forwardchainer/ForwardChainer.cc]: You prefer 3 spaces?

File: opencog/rule-engine/InferenceSCM.h [hunk 2]
```diff
@@ -37,16 +37,19 @@ class InferenceSCM
...
+	 * @param rbs         rule-based system atom
+	 * @param hfocus_set  focus set atoms
 	 *
-	 * @param h target
-	 * @param rbs rule-based system atom
 	 * @return ???
 	 */
-	Handle do_forward_chaining(Handle h, Handle rbs);
-	/**
+    Handle do_forward_chaining(Handle hsource, Handle rbs, Handle hfocus_set);
```
Comment (williampma) [inline opencog/rule-engine/InferenceSCM.h]: Watch out!

File: opencog/rule-engine/forwardchainer/ForwardChainer.h [hunk 3]
```diff
@@ -42,35 +49,84 @@ class ForwardChainer {
...
+    HandleSeq derive_rules(Handle source, Rule* rule, bool subatomic = false);
+
 public:
 	/**
 	 * Ctor. rbs is a Handle pointing to rule-based system.
 	 */
     ForwardChainer(AtomSpace& as, Handle rbs);
-    void do_chain(ForwardChainerCallBack& fcb, Handle hsource =
-            Handle::UNDEFINED);
-    void do_step(ForwardChainerCallBack& fcb);
```
Comment (williampma) [inline opencog/rule-engine/forwardchainer/ForwardChainer.h]: I wonder about the removal of do_step...  it means it is impossible to implement a combined FC/BC (I envision do_step will be used to control one step of FC and one step of BC at a time).

File: opencog/rule-engine/forwardchainer/ForwardChainer.cc [hunk 4]
```diff
@@ -237,3 +244,479 @@ HandleSeq ForwardChainer::get_chaining_result()
...
+        if (contains_atomtype(himplicand, VARIABLE_NODE)) {
+            Handle himplicant = Substitutor::substitute(blptr->get_body(), vgmap);
+
+            //Assuming himplicant's set of variables are superset for himplicand's,
+            //generate varlist from himplicant.
+            Handle hvarlist = as.add_atom(gen_sub_varlist(
+                    himplicant, LinkCast(hrule)->getOutgoingSet()[0]));
+            Handle hderived_rule = as.add_atom(Handle(createBindLink(HandleSeq {
+                    hvarlist, himplicant, himplicand })));
+            derived_rules.push_back(hderived_rule);
```
Comment (williampma) [inline opencog/rule-engine/forwardchainer/ForwardChainer.cc]: Sooner or later you need to use the standardized apart version of the rule to make sure the non-var VariableNode will be uniquely generated.

File: opencog/rule-engine/InferenceSCM.h (hunk 2 above)
Comment (misgeatgit) [inline opencog/rule-engine/InferenceSCM.h]: of what?

Comment (williampma) [inline opencog/rule-engine/InferenceSCM.h]: Just the spaces~

File: opencog/rule-engine/forwardchainer/ForwardChainer.h (hunk 3 above)
Comment (misgeatgit) [inline opencog/rule-engine/forwardchainer/ForwardChainer.h]: OK, I assume this is a feature request.

File: opencog/rule-engine/forwardchainer/ForwardChainer.cc (hunk 4 above)
Comment (misgeatgit) [inline opencog/rule-engine/forwardchainer/ForwardChainer.cc]: I will take that.

Comment (AmeBel) [inline opencog/rule-engine/forwardchainer/ForwardChainer.cc]: what is a "standardized apart version of the rule"?

Comment (williampma) [inline opencog/rule-engine/forwardchainer/ForwardChainer.cc]: A version of the rule with no variable naming conflicts.

**Comment by misgeatgit** [comment]
@williampma  I want to merge this so that @AmeBel  and @leungmanhin  could use it.

**Comment by williampma** [comment]
You need to update your branch.  It's unmergeable right now.

**Comment by williampma** [comment]
Oh, please fix existing usages of `cog-fc` is some future PR.  Thanks.

**Comment by misgeatgit** [comment]
OK.

**Pull request merged**